File: cotsSource/org.apache.qpid/python/qpid/datatypes.py
#
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
#

import threading, struct, datetime, time
from exceptions import Timeout

class Struct:

  def __init__(self, _type, *args, **kwargs):
    if len(args) > len(_type.fields):
      raise TypeError("%s() takes at most %s arguments (%s given)" %
                      (_type.name, len(_type.fields), len(args)))

    self._type = _type

    idx = 0
    for field in _type.fields:
      if idx < len(args):
        arg = args[idx]
        if kwargs.has_key(field.name):
          raise TypeError("%s() got multiple values for keyword argument '%s'" %
                          (_type.name, field.name))
      elif kwargs.has_key(field.name):
        arg = kwargs.pop(field.name)
      else:
        arg = field.default()
      setattr(self, field.name, arg)
      idx += 1

    if kwargs:
      unexpected = kwargs.keys()[0]
      raise TypeError("%s() got an unexpected keyword argument '%s'" %
                      (_type.name, unexpected))

  def __getitem__(self, name):
    return getattr(self, name)

  def __setitem__(self, name, value):
    if not hasattr(self, name):
      raise AttributeError("'%s' object has no attribute '%s'" %
                           (self._type.name, name))
    setattr(self, name, value)

  def __repr__(self):
    fields = []
    for f in self._type.fields:
      v = self[f.name]
      if f.type.is_present(v):
        fields.append("%s=%r" % (f.name, v))
    return "%s(%s)" % (self._type.name, ", ".join(fields))

class Message:

  def __init__(self, *args):
    if args:
      self.body = args[-1]
    else:
      self.body = None
    if len(args) > 1:
      self.headers = list(args[:-1])
    else:
      self.headers = None
    self.id = None

  def has(self, name):
    return self.get(name) != None

  def get(self, name):
    if self.headers:
      for h in self.headers:
        if h.NAME == name:
          return h
    return None

  def set(self, header):
    if self.headers is None:
      self.headers = []
    idx = 0
    while idx < len(self.headers):
      if self.headers[idx].NAME == header.NAME:
        self.headers[idx] = header
        return
      idx += 1
    self.headers.append(header)

  def clear(self, name):
    idx = 0
    while idx < len(self.headers):
      if self.headers[idx].NAME == name:
        del self.headers[idx]
        return
      idx += 1

  def __repr__(self):
    args = []
    if self.headers:
      args.extend(map(repr, self.headers))
    if self.body:
      args.append(repr(self.body))
    if self.id is not None:
      args.append("id=%s" % self.id)
    return "Message(%s)" % ", ".join(args)

def serial(o):
  if isinstance(o, Serial):
    return o
  else:
    return Serial(o)

class Serial:

  def __init__(self, value):
    self.value = value & 0xFFFFFFFFL

  def __hash__(self):
    return hash(self.value)

  def __cmp__(self, other):
    if other.__class__ not in (int, long, Serial):
      return 1

    other = serial(other)

    delta = (self.value - other.value) & 0xFFFFFFFFL
    neg = delta & 0x80000000L
    mag = delta & 0x7FFFFFFF

    if neg:
      return -mag
    else:
      return mag

  def __add__(self, other):
    return Serial(self.value + other)

  def __sub__(self, other):
    if isinstance(other, Serial):
      return self.value - other.value
    else:
      return Serial(self.value - other)

  def __repr__(self):
    return "serial(%s)" % self.value

  def __str__(self):
    return str(self.value)

class Range:

  def __init__(self, lower, upper = None):
    self.lower = serial(lower)
    if upper is None:
      self.upper = self.lower
    else:
      self.upper = serial(upper)

  def __contains__(self, n):
    return self.lower <= n and n <= self.upper

  def __iter__(self):
    i = self.lower
    while i <= self.upper:
      yield i
      i += 1

  def touches(self, r):
    # XXX: are we doing more checks than we need?
    return (self.lower - 1 in r or
            self.upper + 1 in r or
            r.lower - 1 in self or
            r.upper + 1 in self or
            self.lower in r or
            self.upper in r or
            r.lower in self or
            r.upper in self)

  def span(self, r):
    return Range(min(self.lower, r.lower), max(self.upper, r.upper))

  def intersect(self, r):
    lower = max(self.lower, r.lower)
    upper = min(self.upper, r.upper)
    if lower > upper:
      return None
    else:
      return Range(lower, upper)

  def __repr__(self):
    return "%s-%s" % (self.lower, self.upper)

class RangedSet:

  def __init__(self, *args):
    self.ranges = []
    for n in args:
      self.add(n)

  def __contains__(self, n):
    for r in self.ranges:
      if n in r:
        return True
    return False

  def add_range(self, range):
    idx = 0
    while idx < len(self.ranges):
      r = self.ranges[idx]
      if range.touches(r):
        del self.ranges[idx]
        range = range.span(r)
      elif range.upper < r.lower:
        self.ranges.insert(idx, range)
        return
      else:
        idx += 1
    self.ranges.append(range)

  def add(self, lower, upper = None):
    self.add_range(Range(lower, upper))

  def empty(self):
    for r in self.ranges:
      if r.lower <= r.upper:
        return False
    return True

  def max(self):
    if self.ranges:
      return self.ranges[-1].upper
    else:
      return None

  def min(self):
    if self.ranges:
      return self.ranges[0].lower
    else:
      return None

  def __iter__(self):
    return iter(self.ranges)

  def __repr__(self):
    return str(self.ranges)

class Future:
  def __init__(self, initial=None, exception=Exception):
    self.value = initial
    self._error = None
    self._set = threading.Event()
    self.exception = exception

  def error(self, error):
    self._error = error
    self._set.set()

  def set(self, value):
    self.value = value
    self._set.set()

  def get(self, timeout=None):
    self._set.wait(timeout)
    if self._set.isSet():
      if self._error != None:
        raise self.exception(self._error)
      return self.value
    else:
      raise Timeout()

  def is_set(self):
    return self._set.isSet()

try:
  import uuid
  def random_uuid():
    return uuid.uuid4().get_bytes()
except ImportError:
  import random
  def random_uuid():
    bytes = [random.randint(0, 255) for i in xrange(16)]

    # From RFC4122, the version bits are set to 0100
    bytes[7] &= 0x0F
    bytes[7] |= 0x40

    # From RFC4122, the top two bits of byte 8 get set to 01
    bytes[8] &= 0x3F
    bytes[8] |= 0x80
    return "".join(map(chr, bytes))

def uuid4():
  return UUID(random_uuid())

class UUID:

  def __init__(self, bytes):
    self.bytes = bytes

  def __cmp__(self, other):
    if isinstance(other, UUID):
      return cmp(self.bytes, other.bytes)
    else:
      return -1

  def __str__(self):
    return "%08x-%04x-%04x-%04x-%04x%08x" % struct.unpack("!LHHHHL", self.bytes)

  def __repr__(self):
    return "UUID(%r)" % str(self)

  def __hash__(self):
    return self.bytes.__hash__()

class timestamp(float):

  def __new__(cls, obj=None):
    if obj is None:
      obj = time.time()
    elif isinstance(obj, datetime.datetime):
      obj = time.mktime(obj.timetuple()) + 1e-6 * obj.microsecond
    return super(timestamp, cls).__new__(cls, obj)

  def datetime(self):
    return datetime.datetime.fromtimestamp(self)

  def __add__(self, other):
    if isinstance(other, datetime.timedelta):
      return timestamp(self.datetime() + other)
    else:
      return timestamp(float(self) + other)

  def __sub__(self, other):
    if isinstance(other, datetime.timedelta):
      return timestamp(self.datetime() - other)
    else:
      return timestamp(float(self) - other)

  def __radd__(self, other):
    if isinstance(other, datetime.timedelta):
      return timestamp(self.datetime() + other)
    else:
      return timestamp(other + float(self))

  def __rsub__(self, other):
    if isinstance(other, datetime.timedelta):
      return timestamp(self.datetime() - other)
    else:
      return timestamp(other - float(self))

  def __neg__(self):
    return timestamp(-float(self))

  def __pos__(self):
    return self

  def __abs__(self):
    return timestamp(abs(float(self)))

  def __repr__(self):
    return "timestamp(%r)" % float(self)

File: cotsSource/org.apache.qpid/ruby/lib/qpid/datatypes.rb
#
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
#

module Qpid

  def self.struct(type, *args)
    # FIXME: This is fragile; the last arg could be a hash,
    # without being hte keywords
    kwargs = {}
    kwargs = args.pop if args.any? && args[-1].is_a?(Hash)

    if args.size > type.fields.size
      raise TypeError,
      "%s() takes at most %d arguments (%d given)" %
        [type.name, type.fields.size, args.size]
    end

    attrs = type.fields.inject({}) do |attrs, field|
      if args.any?
        attrs[field.name] = args.shift
        if kwargs.key?(field.name)
          raise TypeError,
          "%s() got multiple values for keyword argument '%s'" %
            [type.name, field.name]
        end
      elsif kwargs.key?(field.name)
        attrs[field.name] = kwargs.delete(field.name)
      else
        attrs[field.name] = field.default
      end
      attrs
    end

    unless kwargs.empty?
      unexpected = kwargs.keys[0]
      raise TypeError,
      "%s() got an unexpected keyword argument '%s'" %
        [type.name, unexpected]
    end

    attrs[:st_type] = type
    attrs[:id] = nil

    name = "Qpid_" + type.name.to_s.capitalize
    unless ::Struct.const_defined?(name)
      vars = type.fields.collect { |f| f.name } << :st_type << :id
      ::Struct.new(name, *vars)
    end
    st = ::Struct.const_get(name)

    result = st.new
    attrs.each { |k, v| result[k] = v }
    return result
  end

  class Message

    attr_accessor :headers, :body, :id

    def initialize(*args)
      @body = nil
      @headers = nil

      @body = args.pop unless args.empty?
      @headers = args unless args.empty?

      @id = nil
    end

    def has(name)
      return ! get(name).nil?
    end

    def get(name)
      if @headers
        name = name.to_sym
        @headers.find { |h| h.st_type.name == name }
      end
    end

    def set(header)
      @headers ||= []
      if h = @headers.find { |h| h.st_type == header.st_type }
        ind = @headers.index(h)
        @headers[ind] = header
      else
        @headers << header
      end
    end

    def clear(name)
      if @headers
        name = name.to_sym
        @headers.delete_if { |h| h.st_type.name == name }
      end
    end

    # FIXME: Not sure what to do here
    # Ruby doesn't have a notion of a evaluable string representation
    # def __repr__(self):
    #     args = []
    #     if self.headers:
    #       args.extend(map(repr, self.headers))
    #     if self.body:
    #       args.append(repr(self.body))
    #     if self.id is not None:
    #       args.append("id=%s" % self.id)
    #     return "Message(%s)" % ", ".join(args)
    #   end
  end

  class ::Object

    def to_serial
      Qpid::Serial.new(self)
    end
  end

  class Serial

    include Comparable

    attr_accessor :value

    def initialize(value)
      @value = value & 0xFFFFFFFF
    end

    def hash
      @value.hash
    end

    def to_serial
      self
    end

    def eql?(other)
      other = other.to_serial
      value.eql?(other.value)
    end

    def <=>(other)
      return 1 if other.nil?

      other = other.to_serial

      delta = (value - other.value) & 0xFFFFFFFF
      neg = delta & 0x80000000
      mag = delta & 0x7FFFFFFF

      return (neg>0) ? -mag : mag
    end

    def +(other)
      result = other.to_serial
      result.value += value
      return result
    end

    def -(other)
      result = other.to_serial
      result.value = value - result.value
      return result
    end

    def succ
      Serial.new(value + 1)
    end

    # FIXME: Not sure what to do here
    # Ruby doesn't have a notion of a evaluable string representation
    # def __repr__(self):
    #         return "serial(%s)" % self.value
    # end

    def to_s
      value.to_s
    end

  end

  # The Python class datatypes.Range is emulated by the standard
  # Range class with a few additions
  class ::Range

    alias :lower :begin
    alias :upper :end

    def touches(r)
      # XXX: are we doing more checks than we need?
      return (r.include?(lower - 1) ||
              r.include?(upper + 1) ||
              include?(r.lower - 1) ||
              include?(r.upper + 1) ||
              r.include?(lower)     ||
              r.include?(upper)     ||
              include?(r.lower)     ||
              include?(r.upper))
    end

    def span(r)
      Range.new([lower, r.lower].min, [upper, r.upper].max)
    end

    def intersect(r)
      l = [lower, r.lower].max
      u = [upper, r.upper].min
      return l > u ? nil : Range.new(l, u)
    end

  end

  class RangedSet

    include Enumerable

    attr_accessor :ranges

    def initialize(*args)
      @ranges = []
      args.each { |n| add(n) }
    end

    def each(&block)
      ranges.each { |r| yield(r) }
    end

    def include?(n)
      if (n.is_a?(Range))
        super(n)
      else
        ranges.find { |r| r.include?(n) }
      end
    end

    def add_range(range)
      ranges.delete_if do |r|
        if range.touches(r)
          range = range.span(r)
          true
        else
          false
        end
      end
      ranges << range
    end

    def add(lower, upper = nil)
      upper = lower if upper.nil?
      add_range(Range.new(lower, upper))
    end

    def to_s
      repr = ranges.sort { |a,b| b.lower <=> a.lower }.
        map { |r| r.to_s }.join(",")
      "<RangedSet: {#{repr}}"
    end
  end

  class Future
    def initialize(initial=nil, exception=Exception)
      @value = initial
      @error = nil
      @set = Util::Event.new
      @exception = exception
    end

    def error(error)
      @error = error
      @set.set
    end

    def set(value)
      @value = value
      @set.set
    end

    def get(timeout=nil)
      @set.wait(timeout)
      unless @error.nil?
        raise @exception.new(@error)
      end
      @value
    end
  end

  class UUID
    include Comparable

    attr_accessor :bytes

    def initialize(bytes)
      @bytes = bytes
    end

    def <=>(other)
      if other.respond_to?(:bytes)
        return bytes <=> other.bytes
      else
        raise NotImplementedError
      end
    end

    def to_s
      UUID::format(bytes)
    end

    # FIXME: Not sure what to do here
    # Ruby doesn't have a notion of a evaluable string representation
    #   def __repr__(self):
    #     return "UUID(%r)" % str(self)
    #   end

    def self.random_uuid
      bytes = (1..16).collect { |i| rand(256) }

      # From RFC4122, the version bits are set to 0100
      bytes[7] &= 0x0F
      bytes[7] |= 0x40

      # From RFC4122, the top two bits of byte 8 get set to 01
      bytes[8] &= 0x3F
      bytes[8] |= 0x80
      return bytes.pack("C16")
    end

    def self.uuid4
      UUID.new(random_uuid)
    end

    def self.format(s)
      # Python format !LHHHHL
      # big-endian, ulong, ushort x 4, ulong
      "%08x-%04x-%04x-%04x-%04x%08x" % s.unpack("NnnnnN")
    end
  end
end
